src/components/categories/Create.vue

```diff
@@ -0,0 +1,195 @@
+<script setup lang="ts">
+import { ref } from 'vue'
+import { onBeforeRouteLeave, useRouter } from 'vue-router'
+import { useApi } from '../../composables/useApi'
+import { useAuth } from '../../composables/useAuth'
+import { useSessionToken } from '../../composables/useSessionToken'
+import { Categories } from '../../lib/api/Categories'
+import { CreateCategoryRequest, RunType, SortDirection } from '../../lib/api/data-contracts'
+
+const props = defineProps<{
+	id: number
+}>()
+
+const token = useSessionToken()
+const createError = ref('')
+const router = useRouter()
+
+const categories = new Categories({
+	baseUrl: import.meta.env.VITE_BACKEND_URL
+})
+
+const warnBeforeLeave = ref(true)
+
+const createRequest = ref({
+	name: '',
+	slug: '',
+	info: '',
+	sortDirection: '',
+	type: ''
+})
+
+onBeforeRouteLeave(() => {
+	if (
+		warnBeforeLeave.value &&
+		(createRequest.value.name ||
+			createRequest.value.slug ||
+			createRequest.value.info ||
+			createRequest.value.sortDirection ||
+			createRequest.value.type)
+	) {
+		if (!window.confirm('Do you want to leave? You have unsaved changes.')) {
+			return false
+		}
+	}
+})
+
+function submit() {
+	// Form validation should cover invalid values
+	const request: CreateCategoryRequest = {
+		...createRequest.value,
+		sortDirection: createRequest.value.sortDirection as SortDirection,
+		type: createRequest.value.type as RunType,
+	}
+	useApi(
+		() =>
+			categories.createCategory(props.id, request, useAuth(token.value)),
+		({ data }) => {
+			warnBeforeLeave.value = false
+			router.push({ name: 'categoryView', params: { id: data.id } })
+		},
+		(error) => {
+			createError.value = 'Failed to create: ' + error.status.toString(10)
+		}
+	)
+}
+</script>
+
+<template>
+	<div class="container">
+		<h1>Create Category</h1>
+		<RouterLink
+			class="back-link"
+			:to="{ name: 'leaderboardView', params: { id: props.id } }"
+			title="Changes will not be saved."
+		>
+			&lt; Back
+		</RouterLink>
+		<p v-if="createError" class="error-text">{{ createError }}</p>
+
+		<form @submit.prevent="submit">
+			<p class="info">Fields marked * are mandatory.</p>
+			<table>
+				<tbody>
+					<tr>
+						<th>
+							<label for="name">Name*:</label>
+						</th>
+						<td>
+							<input
+								required
+								v-model="createRequest.name"
+								id="name"
+								class="input"
+							/>
+						</td>
+					</tr>
+					<tr>
+						<th>
+							<label for="slug">Slug*:</label>
+						</th>
+						<td>
+							<input
+								required
+								pattern="[a-zA-Z0-9\-_]{2,80}"
```

Comment:
The backslash and the length specifiers in the regex should be removed.

Reply:
Oh yeah, about that; our original regex in the LB Create Page doesn't actually work. The console complains that the regex is invalid, so we do need the backslash after all. Also on further testing (and I've made sure that there isn't any caching interfering my results); while we do need to keep the length specifier, `minLength` and `maxLength` trigger before `pattern`. My point being that I believe we should keep what I've written here.

Reply:
I just tested it myself and I can confirm that in fact the backslash is entirely necessary. I don't know why I had that wrong. However, The `{2,80}` is entirely redundant and gives a less helpful error message than the minlength and maxlength properties, at least on firefox.

Reply:
If we don't keep the length specifiers in though, we don't pass validation. It's really weird in that the regex overrides the minLength and maxLength attributes if the length specifiers are absent, but vice versa if present. I'm sorry for dragging this conversation out this long, but that's the only way I can get validation to pass.

Reply:
Ah, I see how it works now. Without the length specifiers, the regex only matches one character. The proper way to write it then would have to be `[a-zA-Z0-9\-_]*`, but the length specifiers work too.